ibis/sql/mssql/compiler.py

```diff
@@ -0,0 +1,200 @@
+import pyodbc
+import sqlalchemy as sa
+import sqlalchemy.dialects.mssql as mssql
+
+
+import ibis.sql.alchemy as alch
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.common.exceptions as com
+
+# used for literal translate
+from ibis.sql.alchemy import unary, fixed_arity
+
+
+def raise_unsupported_op_error(translator, expr, *args):
+    msg = "SQLServer backend doesn't support {} operation!"
+    op = expr.op()
+    raise com.UnsupportedOperationError(msg.format(type(op)))
+
+
+def _reduction(func_name, cast_type='int32'):
+    def reduction_compiler(t, expr):
+        arg, where = expr.op().args
+
+        if arg.type().equals(dt.boolean):
+            arg = arg.cast(cast_type)
+
+        func = getattr(sa.func, func_name)
+
+        if where is not None:
+            arg = where.ifelse(arg, None)
+        return func(t.translate(arg))
+
+    return reduction_compiler
+
+
+def _substr(t, expr):
```

Comment:
At the high level, can you explain why mssql compiler here needs to implement custom rules instead of using the shared sql compiler?

Reply:
This method was required and not using the shared sql compiler because of this line `    f = sa.func.substring` since sql server uses a different function name. So this shows that the general shared sql compiler is not general enough. I've seen other sql backends have to do the same thing (e.g. postgresql)